### ISSUE TITLE
Lambda function and transpose operation

### ISSUE THREAD

**Issue by oak-tree** [fchollet/keras#6011]
I have the following function:

    def transpose_dot(vects):
        x, y = vects
        # <x,x> + <y,y> - 2<x,y>
    
        return K.dot(x, K.transpose(y))

When try to evaluate it with `keras` it works 

    x = K.variable(np.array(np_x))
    y = K.variable(np.array(np_x))
    obj = transpose_dot
    objective_output = obj((x, y))
    print('-----------------')
    print (K.eval(objective_output))

result with:

    [[ 1.  1.  1.  2.]
     [ 1.  2.  2.  4.]
     [ 1.  2.  2.  4.]
     [ 2.  4.  4.  8.]

**But**, when try to use it as function for `Lambda` layer it does not work. 
    
    np_x = [[1, 0], [1, 1], [1, 1], [2, 2]]
    features = np.array([np_x])
    test_input = Input(shape=np.array(np_x).shape)
    dot_layer= Lambda(transpose_dot, output_shape=(4,4))([test_input, test_input])
    x = Model(inputs=test_input, outputs=dot_layer)
    x.predict(features, batch_size=1)

Result with 

    self.fn() if output_subset is None else\
    ValueError: Shape mismatch: x has 2 cols (and 4 rows) but y has 4 rows (and 2 cols)
    Apply node that caused the error: Dot22(Reshape{2}.0, Reshape{2}.0)
    Toposort index: 11
    Inputs types: [TensorType(float32, matrix), TensorType(float32, matrix)]
    Inputs shapes: [(4, 2), (4, 2)]
    Inputs strides: [(8, 4), (8, 4)]
    Inputs values: ['not shown', 'not shown']
    Outputs clients: [[Reshape{4}(Dot22.0, MakeVector{dtype='int64'}.0)]]
 

Any idea what I'm missing here?

P.s 
[Stackoverflow link provide as well ](http://stackoverflow.com/questions/43049984/keras-lambda-function-dot-product-mistmatch)

**Comment by unrealwill** [comment]
Hello, 
You are probably missing the fact that using Input( shape=np.array(np_x).shape) silently adds a batching dimension on axis 0.

**Comment by oak-tree** [comment]
@unrealwill 
Thanks, I confused between the batch size and the actual input. I have change it to be 



```
def distance_matrix_part1(vects):
    x, y = vects
    # <x,x> + <y,y> - 2<x,y>
    #return K.transpose(x)
    #
    return K.expand_dims(K.dot(K.transpose(y),x ),axis=0)
```

   ```
 np_x = [[1, 0], [1, 1], [1, 1], [2, 2]]
    features = np.array([np_x])
    test_input = Input(shape=(1,2))
    distance_layer = Lambda(distance_matrix_part1,output_shape=(4,4))([test_input, test_input])
    x = Model(inputs=test_input, outputs=distance_layer)
    x.compile(optimizer="rmsprop",loss='mean_squared_error')
    x.evaluate(features.transpose((1, 0, 2)),y_pred, batch_size=4)
 
```

This actually does run the method. But it complains about the output shape  of Lambda function. The dot function produce a matrix 4x4 based on the input. So inorder to fit the right dim I added `expand_dims` function. But I get the following error 

> ValueError: Error when checking model target: expected lambda_1 to have 3 dimensions, but got array with shape (4, 4)

Which seems like the `expand_dims` function does not add another 1 dim

Any ideas?

**Comment by oak-tree** [comment]
Ok, I got this part working, Transpose actully the same mistake. Because this is a `batch` I need to `reshape` it 

`K.expand_dims(K.dot(K.reshape(x,(x_shape[0],x_shape[1])), K.reshape(y,(y_shape[1],y_shape[0]))),0) `

But because I'm working on the entire batch I'm changing the `batch` dim. But `keras` does not let me to do that.

`"ValueError: Input arrays should have the same number of samples as target arrays. Found 4 input samples and 1 target samples."`

Is it possible to change output of layer including the number of samples ?
